File: src/FileLogger.Old/FileLoggerProvider.cs
using System.Collections.Concurrent;
using System.Runtime.Versioning;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace FileLogger.Old;

[UnsupportedOSPlatform("browser")]
[ProviderAlias("File")]
public sealed class FileLoggerProvider : ILoggerProvider
{
    private readonly ConcurrentDictionary<string, FileLogger> _loggers =
        new(StringComparer.OrdinalIgnoreCase);

    private readonly LogWriter _logWriter;
    private readonly IDisposable? _onChangeToken;
    private FileLoggerConfiguration _currentConfig;

    public FileLoggerProvider(IOptionsMonitor<FileLoggerConfiguration> config, IHostEnvironment hostEnvironment)
    {
        _currentConfig = config.CurrentValue;
        _onChangeToken = config.OnChange(updatedConfig => _currentConfig = updatedConfig);
        _logWriter = new LogWriter(GetCurrentConfig, hostEnvironment);
    }

    public ILogger CreateLogger(string categoryName)
    {
        return _loggers.GetOrAdd(categoryName, name => new FileLogger(name, GetCurrentConfig, _logWriter));
    }

    public void Dispose()
    {
        _loggers.Clear();
        _onChangeToken?.Dispose();
        _logWriter.Dispose();
    }

    private FileLoggerConfiguration GetCurrentConfig()
    {
        return _currentConfig;
    }
}

File: src/asemin.FileLogger/FileLoggerProvider.fs
namespace asemin.FileLogger

open System.Collections.Concurrent
open System.Runtime.Versioning
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Microsoft.Extensions.Options
open asemin.FileLogger.ScopeProvider

[<UnsupportedOSPlatform("browser")>]
[<ProviderAlias("File")>]
type FileLoggerProvider
    (options: IOptionsMonitor<FileLoggerConfig>, timeProvider: ITimeProvider, environment: IHostEnvironment) =
    let controller = new LogController()

    let mapConfig =
        LoggerConfiguration.mapDto environment.ContentRootPath environment.ApplicationName

    do options.CurrentValue |> mapConfig |> controller.UpdateConfig

    let changeSubscription =
        options.OnChange(fun configDtos -> configDtos |> mapConfig |> controller.UpdateConfig)

    let loggers = ConcurrentDictionary<string, FileLogger>()

    let mutable isDisposed = false

    let mutable currentScopeProvider: IExternalScopeProvider =
        NullExternalScopeProvider.Instance

    interface ILoggerProvider with
        member this.CreateLogger(categoryName) =
            RaiseIf.objectDisposed isDisposed this

            loggers.GetOrAdd(
                categoryName,
                (fun name ->
                    FileLogger(categoryName, timeProvider, controller.WriteLog, (fun () -> currentScopeProvider)))
            )

        member this.Dispose() =
            isDisposed <- true
            loggers.Clear()
            changeSubscription.Dispose()
            controller.Dispose()

    interface ISupportExternalScope with
        member this.SetScopeProvider scopeProvider = currentScopeProvider <- scopeProvider
